### PR TITLE
Air Quality Data

### PR DESCRIPTION
**Pull request by Sajeelzafar**
### In this project, the following tasks are being performed:

- Fetching countries from rest countries API.
- Once the country is selected, cities are fetched from the API and loaded.
- Once the city is selected, geocoding API is used to fetch its respective latitude and longitude.
- The coordinates are then passed to air-pollution API to get the data of air.
- The data is displayed in a table.
- The routers are used to navigate through the different stages.
- Tests are performed.
- Readme is updated.

### CODE REVIEW

File: src/Redux/reducers/countries.js [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+export const filteredCountries = (data = '') => (
+  {
+    type: 'FILTER',
+    payload: data,
+  }
+);
+
+const countryReducer = (state = [], action) => {
+  switch (action.type) {
+    case 'FETCH/fulfilled': {
```
Comment (FrederickMih) [inline src/Redux/reducers/countries.js]: - [**Optional**] Hi @Sajeelzafar, so far, all your action types have been strings. Though your application still works, however, in larger applications, using strings for action types can make it harder to track bugs :bug:  We can avoid these types of errors in the future by saving our action types using constants.

File: src/Components/Home.js [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+          variant="secondary"
+          type="button"
+          onClick={async () => {
+            await dispatch(fetchCountries());
+            dispatch(filteredCountries(countryName));
+          }}
+        >
+          Search
+        </Button>
+      </div>
```
Comment (FrederickMih) [inline src/Components/Home.js:32]: - [x] Your filter work perfectly :+1: however, we are asked to use the **Filter** stateless component to filter data retrieved from the API. Let's kindly have a Filter stateless component on its own to carry out the filtering.

File: README.md [hunk 3]
```diff
@@ -1 +1,148 @@
...
+- **[Interactive buttons to load specific cities and air data when pressed]**
+- **[Styling done using React Bootstrap]**
+
+<p align="right">(<a href="#readme-top">back to top</a>)</p>
+
+## 🚀 Live Demo <a name="live-demo"></a>
+
+> Link to deployed project is:
+
+- [Live Demo Link](https://space-x1234.netlify.app/)
```
Comment (FrederickMih) [inline README.md]: - [x] Your project should be deployed properly to Heroku or Netify. The Home page and the detail page should work as expected. A link to the deployed project should be present in your README. This link does not contain the expected project. Kindly fix :wrench: . This requirement if essential to the approval of this project.

File: src/Components/Home.js [hunk 4]
```diff
@@ -0,0 +1,68 @@
+import { useEffect } from 'react';
+import { Button, Card, Container } from 'react-bootstrap';
+import { useDispatch, useSelector } from 'react-redux';
+import { v4 as uuidv4 } from 'uuid';
+import { NavLink } from 'react-router-dom';
+import { fetchCountries, filteredCountries } from '../Redux/reducers/countries';
+
+const Home = () => {
```
Comment (FrederickMih) [inline src/Components/Home.js:8]: - [x] We are expected to follow the design guidelines for this project including typography, and layout. Once more try to check out the screenshots below and make the changes.

| Yours | Guidelines |
| --- | --- |
| Screenshot from 2022-12-02 19-24-28 | Screenshot from 2022-12-02 19-27-09 |

- Kindly note that having a nice design increases your chances for a job as recruiters will be checking your app.

File: src/tests/App.test.js [hunk 5]
```diff
@@ -0,0 +1,24 @@
+import React from 'react';
+import { render } from '@testing-library/react';
+import '@testing-library/jest-dom';
+import { BrowserRouter as Router, Route, Routes } from 'react-router-dom';
+import { Provider } from 'react-redux';
+import store from '../Redux/configureStore';
+import City from '../Components/City';
+import Home from '../Components/Home';
+import Pollution from '../Components/Pollution';
```
Comment (FrederickMih) [inline src/tests/App.test.js:9]: - [x] Please we need more tests. Based on the requirements of this project, the app should have unit tests for several components of the app and some pure functions in Redux code. Also, the app should have integration tests using React testing library.

**Review by FrederickMih — CHANGES_REQUESTED** [review]
## Hi :raising_hand: @Sajeelzafar , 

Fantastic job so far!
You have done an amazing job :+1: on this project.
There are some issues that you still need to work on to prepare your project for the final evaluation but you are almost there!



Screenshot from 2022-10-20 09-38-51


## Highlights :green_circle: 

- App has the required number of pages :heavy_check_mark: 
- Redux was used for state management :+1: 
- PR is descriptive :+1: 
- Linters checks are OK :+1: 

## Suggested changes

- [x] In your presentation, you didn't highlight the piece of code you are proud of as stipulated in the requirements.

_Check the comments under the review._

You can use as many of my suggestions as you want. If there is anything you would like to skip - feel free to do that. However, I strongly recommend you to take them into account as they can make your code better._



Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

File: README.md (hunk 3 above)
Comment (Sajeelzafar) [reply README.md]: Thank you for your time, I have updated the README.md with the correct link.

File: src/tests/App.test.js (hunk 5 above)
Comment (Sajeelzafar) [reply src/tests/App.test.js:9]: Thank you for the feedback @FrederickMih, I have added additional tests for components. Thank you for making the code better.

File: src/Redux/reducers/countries.js (hunk 1 above)
Comment (Sajeelzafar) [reply src/Redux/reducers/countries.js]: Thank you @FrederickMih for your suggestion. I have renamed every action to constant and will try adopting it in the future as well.

File: src/Components/Home.js (hunk 4 above)
Comment (Sajeelzafar) [reply src/Components/Home.js:8]: Hi @FrederickMih, thank you for pointing that out. However, the theme designs for both apps are very much different. I have tried altering my design to match the given design. I hope it resembles better now.

**Review by KDlamini — APPROVED** [review]
### Hi @Sajeelzafar 🙋‍♀️,  

Excellent work! 💯 
You have done a great job implementing all the required changes. 🎊 🥳 
Your feature is complete! There is nothing else to say other than... it's time to merge it. :shipit: 
Congratulations!!!

## HIGHLIGHTS 🌟 

✔️ Professional pull request
✔️ React components, events, and state management well-implemented
✔️ Tests are added
✔️ Application works as expected
✔️ No linter errors

# Status: Approved ✔️ 

LeonardoInGatsby

Cheers 🥂  and Happy coding!!! 👯 
Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Remember to tag me @KDlamini  in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think the code review was unfair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**